File: Regular_Expression_Matching.cpp
/*10. Regular Expression Matching  QuestionEditorial Solution  My Submissions
Total Accepted: 101002 Total Submissions: 438570 Difficulty: Hard
Implement regular expression matching with support for '.' and '*'.

'.' Matches any single character.
'*' Matches zero or more of the preceding element.

The matching should cover the entire input string (not partial).

The function prototype should be:
bool isMatch(const char *s, const char *p)

Some examples:
isMatch("aa","a") → false
isMatch("aa","aa") → true
isMatch("aaa","aa") → false
isMatch("aa", "a*") → true
isMatch("aa", ".*") → true
isMatch("ab", ".*") → true
isMatch("aab", "c*a*b") → true
Hide Company Tags Google Uber Airbnb Facebook Twitter
Hide Tags Dynamic Programming Backtracking String
Hide Similar Problems (H) Wildcard Matching
*/



/*DP dynamic programming
use dp[i][j] to check if two string is match the pattern
first initialize false, except dp[0][0]=true
dp[i][j]=true, if s[0...i-1] matches p[0...j-1]
i is control s, j is control p
P[i][j] = P[i - 1][j - 1], if p[j - 1] != '*' && (s[i - 1] == p[j - 1] || p[j - 1] == '.');
P[i][j] = P[i][j - 2], if p[j - 1] == '*' and the pattern repeats for 0 times;
P[i][j] = P[i - 1][j] && (s[i - 1] == p[j - 2] || p[j - 2] == '.'), if p[j - 1] == '*' and the pattern repeats for at least 1 times.

if p[j - 1] != '*'
   f[i][j] = f[i - 1][j - 1] && s[i - 1] == p[j - 1]
if p[j - 1] == '*', denote p[j - 2] with x
   f[i][j] is true iff any of the following is true
   1) "x*" repeats 0 time and matches empty: f[i][j - 2]
   2) "x*" repeats >= 1 times and matches "x*x": s[i - 1] == x && f[i - 1][j]
'.' matches any single character
return dp[m][n]*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
	bool isMatch(string s, string p) {
		int m = s.size();
		int n = p.size();
		vector<vector<bool>> dp(m + 1, vector<bool>(n + 1, false));
		dp[0][0] = true;
		for (int i = 0; i <= m; ++i) {
			for (int j = 1; j <= n; ++j) {
				if (p[j - 1] == '*') {
					dp[i][j] = dp[i][j - 2] || (i > 0 && dp[i - 1][j] && (s[i - 1] == p[j - 2] || p[j - 2] == '.') );
				} else {
					dp[i][j] = i > 0 && dp[i - 1][j - 1] && (s[i - 1] == p[j - 1] || p[j - 1] == '.');
				}
			}
		}
		return dp[m][n];
	}
};




/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
	public boolean isMatch(String s, String p) {
		int m = s.length();
		int n = p.length();
		boolean[][] dp = new boolean[m + 1][n + 1];
		dp[0][0] = true; // If s and p are "", isMathch() returns true;
		for (int i = 0; i <= m; ++i) {
			for (int j = 1; j <= n; ++j) {//start  from 1
				if (p.charAt(j - 1) == '*') {
					// Two situations:
					// (1) dp[i][j-2] is true, and there is 0 preceding element of '*';
					// (2) The last character of s should match the preceding element of '*';
					//     And, dp[i-1][j] should be true;
					dp[i][j] = dp[i][j - 2]
					           || (i > 0 && dp[i - 1][j]
					               && (s.charAt(i - 1) == p.charAt(j - 2) || p.charAt(j - 2) == '.'));
				} else {
					// The last character of s and p should match;
					// And, dp[i-1][j-1] is true;
					dp[i][j] = i > 0 && dp[i - 1][j - 1] &&
					           (s.charAt(i - 1) == p.charAt(j - 1) || p.charAt(j - 1) == '.');
				}
			}
		}
		return dp[m][n];
	}
}

File: Regular_Expression_Matching.swift
/*
Given an input string (s) and a pattern (p), implement regular expression matching with support for '.' and '*' where: 

'.' Matches any single character.​​​​
'*' Matches zero or more of the preceding element.
The matching should cover the entire input string (not partial).

 

Example 1:

Input: s = "aa", p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".
Example 2:

Input: s = "aa", p = "a*"
Output: true
Explanation: '*' means zero or more of the preceding element, 'a'. Therefore, by repeating 'a' once, it becomes "aa".
Example 3:

Input: s = "ab", p = ".*"
Output: true
Explanation: ".*" means "zero or more (*) of any character (.)".
Example 4:

Input: s = "aab", p = "c*a*b"
Output: true
Explanation: c can be repeated 0 times, a can be repeated 1 time. Therefore, it matches "aab".
Example 5:

Input: s = "mississippi", p = "mis*is*p*."
Output: false
 

Constraints:

0 <= s.length <= 20
0 <= p.length <= 30
s contains only lowercase English letters.
p contains only lowercase English letters, '.', and '*'.
It is guaranteed for each appearance of the character '*', there will be a previous valid character to match.
*/

/*
Solution 3:
DP bottom up

Time Complexity: O(countS * countP)
Space Complexity: O(countS * countP)
*/
class Solution {
    func isMatch(_ s: String, _ p: String) -> Bool {
        if s.isEmpty, p.isEmpty { return true }
        
        var s = Array(s)
        var p = Array(p)
        
        var countS = s.count
        var countP = p.count
        
        var dp: [[Bool]] = Array(repeating: Array(repeating: false, count: countP+1), 
                       count: countS+1)
        
        dp[countS][countP] = true
        
        for indexS in stride(from: countS, through: 0, by: -1) {
            for indexP in stride(from: countP-1, through: 0, by: -1) {
                // print(indexS, indexP, countS, countP)
                let isFirstMatch = (indexS < countS 
                                    && (s[indexS] == p[indexP] || p[indexP] == "."))
                // print(isFirstMatch)
                if indexP < countP-1, p[indexP+1] == "*" {
                    dp[indexS][indexP] = dp[indexS][indexP+2] 
                    || (isFirstMatch && dp[indexS+1][indexP])
                } else {
                    dp[indexS][indexP] = isFirstMatch && dp[indexS+1][indexP+1]
                }
            }
        }
        
        return dp[0][0]
    }
}

/*
Solution 2:
DP top down

use extra dp[indexS, indexP] to store already checked result

Time Complexity: O(countS * countP)
Space Complexity: O(countS * countP)
*/
class Solution {
    func isMatch(_ s: String, _ p: String) -> Bool {
        if s.isEmpty, p.isEmpty { return true }
        
        var s = Array(s)
        var p = Array(p)
        
        var countS = s.count
        var countP = p.count
        
        // 0 means not checked 
        // 1 means false
        // 2 means true
        var dp: [[Int]] = Array(repeating: Array(repeating: 0, count: countP+1), 
                       count: countS+1)
        
        func check(_ indexS: Int, _ indexP: Int) -> Bool {
            if dp[indexS][indexP] != 0 {
                return dp[indexS][indexP] == 2
            }
            
            var res = false
            
            if indexP == countP {
                res = (indexS == countS)
            } else {
                var isFirstMatch = (indexS < countS 
                                    && (s[indexS] == p[indexP] || p[indexP] == "."))
                
                if indexP < countP-1, p[indexP+1] == "*" {
                    res = (check(indexS, indexP+2) 
                           || (isFirstMatch && check(indexS+1, indexP)))
                } else {
                    res = (isFirstMatch && check(indexS+1, indexP+1))
                }
            }
            
            dp[indexS][indexP] = (res == true ? 2 : 1)
            return res
        }
        
        return check(0, 0)
    }
}

/*
Solution 1:
recursive

Time Complexity: O((countS + countP) * 2^(countS + countP/2))
Space Complexity: O(countS^2 + countP^2)
*/
class Solution {
    func isMatch(_ s: String, _ p: String) -> Bool {
        if s.isEmpty, p.isEmpty { return true }
        return _isMatch(Array(s), Array(p), 0, 0, s.count, p.count)
    }
    
    func _isMatch(_ s: [Character], 
                  _ p: [Character],
                  _ indexS: Int, _ indexP: Int, 
                  _ countS: Int, _ countP: Int) -> Bool {
        if indexP == countP { return indexS == countS }
        var isFirstMatch = (indexS < countS 
                            && (s[indexS] == p[indexP] || p[indexP] == "."))
        
        if indexP < countP-1, p[indexP+1] == "*" {
            return _isMatch(s, p, 
                            indexS, indexP+2, 
                            countS, countP)
            || (isFirstMatch && _isMatch(s, p, 
                                         indexS+1, indexP, 
                                         countS, countP))
        } else {
            return isFirstMatch && _isMatch(s, p, 
                                            indexS+1, indexP+1, 
                                            countS, countP)
        }
    }
}